### PR TITLE
bring back v1.1-style transliteration during encoding

### LINKED ISSUE #3526: RStudio Preview fails to parse multibyte characters on non-UTF-8 MBCS Windows
The stable version of RStudio is fine, but with the preview version, I got following errors related to multibyte strings. Note that, `source()`ing and `reprex::reprex()` works fine with the very same code. So, I suspect something is wrong in the preview version of RStudio.

### System details

    RStudio Edition : Desktop
    RStudio Version : 1.2.981
    OS Version      :  Windows 10
    R Version       :  3.5.1
    Locale          : Japanese_Japan.932

### Steps to reproduce the problem

```r
# this works, so multibyte characters are basically OK
data.frame("Φ" = 1)

# But this won't
data.frame(Φ = 1)

# This won't work, too
data.frame(`Φ` = 1)

# Hmm...?
data.frame(`ç`` = 1)

# Hmmmm...??
data.frame(`ç` = 1)
```

### Describe the problem in detail

image

### Describe the behavior you expected

- First three should work
- Forth should be treated as incomplete `` ` ``
- Fifth should work (But I guess I can't stop `ç` is transtated into `c`)

### PR DESCRIPTION
**Pull request by kevinushey**
Closes #3526, #3496.

### CODE REVIEW

**Comment by kevinushey** [comment]
Hmm, it looks like the issue is a bit more subtle: the notion of the active locale is different internally with R and our active R session. It's as though our code is being run with the C locale active, while R is running with a true system locale (e.g. "English"). It's the locale that defines the transliteration rules, so when run in R's context we get transliteration with `::wctomb()`, but in our own context we don't perform transliteration and just fail to re-encode. I suspect this is still an MSVC vs. MinGW issue of some sort.

Another way to fix this, then, would be to communicate R's locale back to RStudio and update it with e.g. `::setlocale()` as needed; not sure if we want to go that route.

**Comment by kevinushey** [comment]
I think this is ready for review.

I'm not sure what the best way to unit test this is, though -- is there a natural way we could exercise console input + output with the R session active? (`.rs.invokeRpc("console_input", "...")` didn't seem to work as I might've hoped)

File: src/cpp/core/StringUtils.cpp [hunk 1]
```diff
@@ -262,29 +262,36 @@ std::string utf8ToSystem(const std::string& str,
...
    {
       LOG_ERROR(LAST_SYSTEM_ERROR());
       return str;
    }
 
    std::ostringstream output;
-   char mbbuf[10];
-   // Only go up to chars - 1 because last char is \0
-   for (int i = 0; i < chars - 1; i++)
+   char buffer[16];
```
Comment (gtritchie) [inline src/cpp/core/StringUtils.cpp]: Where did 16 come from? MSDN and C++ docs say `MB_CUR_MAX` or at worst `MB_LEN_MAX` are appropriate. These style of functions are generally scary; we weren't allowed to use them in MS-Office code at all, replaced with `wctomb_s`.

**Comment by jmcphers** [comment]
> is there a natural way we could exercise console input + output with the R session active?

This is hard to unit test because it's async; we'd need to fire the console input to R and then hang around and wait for output. There's no test harness currently for R output so you'd have to add that to use the R session testing infrastructure.

All that might be worth doing but probably not just for this change.

File: src/cpp/r/session/RStdCallbacks.cpp [hunk 2]
```diff
@@ -445,7 +445,13 @@ void RBusy(int which)
 {
    try
    {
-      s_callbacks.busy(which == 1) ;
+      // synchronize locale whenever R busy state changes
+      // (done relatively eagerly to ensure synchronization
+      // happens on each REPL iteration after user code is run)
+      r::util::synchronizeLocale();
```
Comment (jmcphers) [inline src/cpp/r/session/RStdCallbacks.cpp]: This requires the `base` package to be attached -- can R ever be busy before then?

It feels like this should be hooked up to OnDetectChanges with `ChangeSourceREPL` instead. Does that seem right?

Comment (jmcphers) [reply src/cpp/r/session/RStdCallbacks.cpp]: Okay, fair enough!

**Pull request merged**